### PR TITLE
feat: Add Links Application - MEED-2512 - Meeds-io/MIPs#92

### LINKED ISSUE #92: Links portlet
## Rationale

As a platform admin, it is not possible to add links while it'd be interesting to hilight some links in a page for example in the public page we aim to deliver soon.

### Top User Stories

As a platform admin OR as a space host, I can add links using "Links Portlet" with these characteristics:

**Precision**
* Who can edit it?
A. Platform admin or Publisher (in web-contributors group) for platforms page
B. Space host or space publisher for spaces
-> When doing so, admin or space host remain able to edit links (they are publisher by design). That means this is a delegation the admin can give to other users.
-> That still means that redactors cannot edit links.
* Who can see it?
A. editors (see above)
B. anyone who can access the page (Same access permissions of the current page)

**Links display**
I can choose between 3 display: rows, cards, columns

image

For each display, I can decide:
a. to add header
b. to add a see more
c. to show name
d. to use large icons

Depending what characteristic I add to the display, the preview is updated dynamically:
image

**Links addition**
When I add a link, then I can:
a. name it
b. describe it (to be displayed in a tooltip only)
c. upload an icon (default one is a globe)

Once added, I can decide the order of the links with move up / move down options

image

### Impacts

NA

#### Gamification
NA

#### Notifications
NA

#### Analytics
NA

#### Unified Search
NA

### Security 

The Links portlet content access has to follow the following ACL rules:

**Who can Edit it**

- All Platform administrators
- Publisher in Platform Web Contributors
- (If page in a space) Publisher & Host roles of the space

**Who can Access it**

- Editors (See above)
- Same access permission as the page

### Upgradability

When the `merge` or `overwrite` import mode are applied to pages, the portlet settings has to remain as was set using UI.

### Feature Flags

The feature flag `exo.feature.publisherRolePromotion.enabled` will be removed (introduced by MIP #35) to give full access to the publisher role management in space Members application.

### Security and access

In order to allow reusing the same setting even after restarting a server using sites import mode `merge` and `overwrite`, the settings of the portlet has to be stored in a dedicated Table (we can also use an unstructured data storage like `SettingService` Key/Value, but for better extensibility, we will use database tables to structure the data to store). When the server re-import the page, the Link Portlet settings has to remain. To do this, a portlet preference will be added in portlet definition in XML file to allow referencing the `Link Portlet Settings`. This name is unique, same as a technical identifier and could be reused in different portlet instances in different pages. This complexity of configuration is hidden to end user and will just permit to preserve the settings changed by UI.
If a portlet is added on the fly (by UI by example), the name of the setting will be exactly the same as the portlet technical identifier without having to request user to add a unique name for portlet instance.
The settings has to be accessible through the portlet technical identifier (`name` preference) to retrieve the referenced page and contextual space. The ACL check using page permissions and space will allow retrieving the list of links for current user or deny it. This way, the list of links aren't all time public and could be used in restricted areas/pages.

### Data and persistence

The tables that will be used to store each Links Portlet instance is as following:

271041890-d257a005-7e6a-4af5-840b-7375adb2bffe drawio

The Link Setting Header, link name and link description are translated, thus the `TranslationService` API will be used to persist those information instead of storing it in Links Table.
Each link setting will be added to each portlet instance, thus the `PAGE_REFERENCE` and `SPACE_ID` Table columns will be used to store the page reference and contextual space of the page. By using both information, the Links Service Layer will be able to check on users permissions without relying on REST or WebUI contextual information (ConversationState & currently displayed Site).

### PR DESCRIPTION
**Pull request by boubaker**
This change will introduce a new application for links to display in a page. The edit and access permission of the application will follow the permissions set on the page.

### CODE REVIEW

File: component/core/src/main/resources/db/changelog/links-rdbms.db.changelog-1.0.0.xml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (C) 2003-2021 eXo Platform SAS.
```
Comment (rdenarie) [inline component/core/src/main/resources/db/changelog/links-rdbms.db.changelog-1.0.0.xml:3]: Date are not correct

**Review by rdenarie — COMMENTED** [review]
By testing this portlet Link, I meet a problem.

I add the portlet on a page, then edit it, and see in settings tab the field 'name'. 
Firstly I dont understand what it is, and dont fill it and try to save. It says : 'name is mandatory'. 
So I fill it without knowing exactly what it is. 
Then I remember it is the reference for the LinkService.

So I retest to add the portlet on a page, without going in portlet settings 
I can save without problem. 
When I go back editing the portlet setting, I see a name was generated automatically.

As this parameter must be unique, and is technical, perhaps we should not display it to user, and always generate it automatically. So for user, there is no portlet settings to take care. 

Currently, it is a little confusing.

cc @srenault-meeds

**Comment by boubaker** [comment]
> Currently, it is a little confusing.

Yes, already tracked after internal tests on eXo ACC, will be improved @rdenarie

File: component/core/src/main/resources/db/changelog/links-rdbms.db.changelog-1.0.0.xml (hunk 1 above)
Comment (boubaker) [reply component/core/src/main/resources/db/changelog/links-rdbms.db.changelog-1.0.0.xml:3]: Even the license was attributed to eXo, which is wrong. Fixed, thanks.

**Comment by rdenarie** [comment]
Will it improved in this MIPs ? or after ?

**Comment by boubaker** [comment]
> Will it improved in this MIPs ? or after ?

already fixed by https://github.com/Meeds-io/social/pull/2970/commits/e93b7db8907a428c24ed322c7da1cdae85ee1861, the ACCs will be restarted in few minutes.

**Comment by boubaker** [comment]
> > Will it improved in this MIPs ? or after ?
> 
> already fixed by [e93b7db](https://github.com/Meeds-io/social/commit/e93b7db8907a428c24ed322c7da1cdae85ee1861), the ACCs will be restarted in few minutes.

ACC restarted, I let you confirm resolution @rdenarie

**Review by rdenarie — DISMISSED** [review]

**Review by rdenarie — DISMISSED** [review]

**Comment by boubaker** [comment]
@rdenarie Meeds-io/MIPs#54 merged to develop, can you review this one again please

**Review by rdenarie — APPROVED** [review]

**Pull request merged**